File: frontend/src/api/external-asset-data/os-ngd-data-source-handler.ts
import type { Feature } from "geojson";
import { DataSourceHandler } from "./data-source-handler";
import { AssetSpecification } from "@/hooks/queries/dataset-utils";

type Filters = {
  description?: string | string[];
  type?: string | string[];
  cqlFilter?: string;
};

type Link = {
  href: string;
  rel?: string;
  type?: string;
  title?: string;
};

/**
 * Represents the data structure returned from the OS NGD API.
 */
interface NGDFeatureCollection {
  type: "FeatureCollection";
  links: Link[];
  features: Feature[];
  numberReturned: number;
}

const filterableFields = ["buildinguse", "roadstructure"];

export class OsNgdDataSourceHandler extends DataSourceHandler {
  /**
   * A function to build a query string.
   *
   * @param allFilters a set of filters which can be parameterised as a query string
   * @param filters a list of pre-configured filters (default is an empty list)
   * @returns a query string which can be suffixed to a URL
   */
  protected buildQueryString(
    allFilters: Filters,
    filters: [string, string][] = [],
  ): string {
    for (const [key, value] of Object.entries(allFilters)) {
      if (typeof value === "string" && filterableFields.includes(key)) {
        filters.push([key, value]);
      }
    }

    let filter: string = filters
      .map(
        ([key, val]) =>
          `filter=${encodeURIComponent(key)}='${encodeURIComponent(val)}'`,
      )
      .join("&");

    if (allFilters.cqlFilter) {
      if (filters.length > 0) {
        console.warn(
          "Do not specify other filters when using cqlFilter",
          allFilters,
        );
      }
      filter = `filter=${encodeURIComponent(allFilters.cqlFilter)}`;
    }

    return filter ? `?bbox=${this.locator}&${filter}` : `?bbox=${this.locator}`;
  }

  /**
   * A function to build a list of query strings.
   *
   * @param filters a set of filters which can be parameterised as query strings
   * @returns a list of query strings
   */
  protected buildQueryStrings(filters: Filters): string[] {
    let { description } = filters;
    if (!description || description.length === 0) {
      return [this.buildQueryString(filters)];
    } else {
      description = Array.isArray(description) ? description : [description];

      return description.map((desc) => {
        const allFilters: [string, string][] = [["description", desc]];
        return this.buildQueryString(filters, allFilters);
      });
    }
  }

  /**
   * A function to build a list of OS NGD URLs.
   *
   * @param assetSpecification the specification of the asset
   * @returns
   */
  public buildUrlsForDataSource(
    assetSpecification: AssetSpecification,
  ): string[] {
    const queryStrings: string[] = this.buildQueryStrings(assetSpecification);
    const urls: string[] = [];
    for (const queryString of queryStrings) {
      const url = `/transparent-proxy/os-ngd/features/ngd/ofa/v1/collections/${assetSpecification.collection}/items${queryString}`;
      const count = assetSpecification.expectedCount;
      if (count && count > 0) {
        const numberPages = Math.ceil(count / 100);
        for (let i = 0; i < numberPages; i++) {
          urls.push(`${url}&offset=${i}00`);
        }
      } else {
        urls.push(url);
      }
    }
    return urls;
  }

  /**
   * A transformation function to replace the external address of the NGD API in a URL with the proxy route.
   * @param url the URL to transform
   * @returns the transformed URL
   */
  protected transformUrl(url: string): string {
    const url_replaced_stem = url.replace(
      "https://api.os.uk",
      "/transparent-proxy/os-ngd",
    );
    return url_replaced_stem.split("&key")[0];
  }

  /**
   * A function which fetches all data from the OS NGD API for a given asset specification.
   *
   * @param assetSpecification the specification of the asset
   * @param url the URL to be queried for data
   * @returns a list of mapped GeoJSON-compatible features
   */
  public async fetchDataForAssetSpecification(
    assetSpecification: AssetSpecification,
    url?: string,
  ): Promise<Feature[]> {
    const allFeatures: Feature[] = [];
    const paginate = !url?.includes("offset=");
    while (url !== undefined) {
      const currentUrl = url; // capture the value for this iteration

      const ngdFeatureCollection: NGDFeatureCollection = await this.limit(
        () => currentUrl && this.fetchFromUrlWithRetry(currentUrl),
      );
      const features = this.mapNGDToGeoJSON(
        ngdFeatureCollection,
        assetSpecification,
      );
      allFeatures.push(...features);
      if (paginate) {
        url = this.getNextUrl(ngdFeatureCollection.links);
      } else {
        const onLastPage =
          assetSpecification.expectedCount - this.getOffset(currentUrl) < 100;
        if (onLastPage && ngdFeatureCollection.numberReturned === 100) {
          url = this.getNextUrl(ngdFeatureCollection.links);
        } else {
          break;
        }
      }
    }

    return allFeatures;
  }

  private getNextUrl(links: Link[]): string | undefined {
    const nextUrl = links.find((item) => item.rel === "next")?.href;
    return nextUrl && this.transformUrl(nextUrl);
  }

  private getOffset(url: string): number {
    const urlFragments = url.split("offset=");
    const offsetStr = urlFragments[1];
    return Number(offsetStr);
  }

  /**
   * A mapper function which creates GeoJSON features from the response returned by OS NGD.
   * It also filters out any features returned by OS NGD which do not match the specification of the asset.
   *
   * @param ngdData the response returned from the OS NGD API
   * @param assetSpecification the specification of the asset
   * @returns an array of Features
   */
  public mapNGDToGeoJSON(
    ngdData: NGDFeatureCollection,
    assetSpecification: AssetSpecification,
  ): Feature[] {
    return ngdData.features
      .filter((f) =>
        this.isMatchForAssetSpecificationFilters(assetSpecification, f),
      )
      .map((f) => {
        return {
          ...f,
          properties: {
            ...f.properties,
            name: (f.properties as { name1_text: string }).name1_text,
          },
        };
      });
  }
}

File: backend/vista-python-api/src/api/repository/external/asset_data_source_handlers/os_ngd_data_source_handler.py
# SPDX-License-Identifier: Apache-2.0
# © Crown Copyright 2026. This work has been developed by the National Digital Twin Programme and is legally attributed to the Department for Business and Trade (UK) as the governing entity.

"""Handler for the OS NGD data source."""

import urllib
import warnings
from typing import ClassVar

from django.conf import settings

from api.models.asset import Asset
from api.repository.external.external_asset_mapper import ExternalAssetMapper

from .base import DataSourceHandler


class OsNgdDataSourceHandler(DataSourceHandler):
    """Handler for the OS NGD data source."""

    filterable_fields: ClassVar[list[str]] = ["buildinguse", "roadstructure"]
    os_ngd_response_page_size: ClassVar[int] = 100

    def build_urls_for_data_source(self, asset_specification):
        """Build the URLs for fetching the data per the specification given."""
        query_strings = self._build_query_strings(asset_specification)
        return [
            f"https://api.os.uk/features/ngd/ofa/v1/collections/"
            f"{asset_specification['collection']}/items{query_string}"
            for query_string in query_strings
        ]

    async def fetch_data_for_asset_specification(
        self, asset_specification, url: str
    ) -> list[Asset]:
        """Fetch the OS NGD data per the specification given."""
        fetch_next_page = True
        offset = 0
        assets = []
        while fetch_next_page:
            offset_url = url + f"&offset={offset}"
            response = await self.fetch_from_url_with_retry(offset_url)
            fetch_next_page = (
                response["numberReturned"] == self.os_ngd_response_page_size
                if "numberReturned" in response
                else False
            )
            offset += self.os_ngd_response_page_size
            filtered_assets = []
            for feature in response["features"]:
                mapped_asset = ExternalAssetMapper.map_from_os_ngd(feature, asset_specification)
                if "filters" in asset_specification:
                    if self._is_feature_match_for_asset_specification_filter(
                        asset_specification["filters"], feature
                    ):
                        filtered_assets.append(mapped_asset)
                else:
                    filtered_assets.append(mapped_asset)
            assets.extend(filtered_assets)
        return assets

    def _build_query_string(
        self, all_filters: dict[str, str], filters: list[tuple[str, str]] | None = None
    ) -> str:
        filters = [] if filters is None else filters
        for key, value in all_filters.items():
            if isinstance(value, str) and key in self.filterable_fields:
                filters.append((key, value))

        filter_parts = [
            f"filter={urllib.parse.quote(key)}='{urllib.parse.quote(val)}'" for key, val in filters
        ]
        filter_str = "&".join(filter_parts)

        if all_filters.get("cqlFilter"):
            if filters:
                warnings.warn(
                    f"Do not specify other filters when using cqlFilter: {all_filters}",
                    stacklevel=2,
                )
            filter_str = f"filter={urllib.parse.quote(all_filters['cqlFilter'])}"

        if filter_str:
            return f"?key={settings.OS_NGD_API_KEY}&bbox={self.locator}&{filter_str}"
        return f"?key={settings.OS_NGD_API_KEY}&bbox={self.locator}"

    def _build_query_strings(self, filters: dict[str, str | list[str]]) -> list[str]:
        description = filters.get("description")

        if not description:
            return [self._build_query_string(filters, [])]

        if not isinstance(description, list):
            description = [description]

        query_strings = []
        for desc in description:
            all_filters = [("description", desc)]
            query = self._build_query_string(filters, all_filters)
            query_strings.append(query)
        return query_strings

    def _is_feature_match_for_asset_specification_filter(self, data_filters, feature) -> bool:
        if data_filters:
            matches_all_filters = True
            for data_filter in data_filters:
                if isinstance(data_filter["filterValue"], list):
                    matches_all_filters = self._is_match_for_any_filter_value(feature, data_filter)
                else:
                    matches_all_filters = (
                        "properties" in feature
                        and feature["properties"].get(data_filter["filterName"])
                        == data_filter["filterValue"]
                    )
            return matches_all_filters
        return True

    def _is_match_for_any_filter_value(self, feature, data_filter):
        return any(
            "properties" in feature and feature["properties"].get(data_filter["filterName"]) == prop
            for prop in data_filter["filterValue"]
        )
